Assertion in UnsafeRow "sizeInBytes should be a multiple of 8"

We just tried to upgrade our pipeline to Spark 2.3 and thus frameless 0.6.0 and hit some nasty assertions in sparks UnsafeRow. We have assertions enabled in our codebase. We tried to reproduce the issue in the frameless repo, but found that `-ea` was not provided for the tests.

Here is the PR introducing them in Spark 2.3: https://github.com/apache/spark/pull/18503/files

As it seems, during codegen, spark is getting a serialized row for a case class (using the RecordEncoder) which only contains not nullable fields, so the nullable bitset is completely `0` and the values are directly in the `values` section elements instead of pointing to the variable length section at the end. But the generated code seems to expect the struct, representing the serialized case class (or tuple) in the first field of the given row. So the `UnsafeRow.getStruct` method used and expects a field pointing to the variable length section at field ordinal `0` and thus interprets the 64 bits at this location (which are the actual long value) as a pointer with offset and size each 32 bits).

I don't exactly know where the error is. If the `RecordEncoder`, which seems to be at play here is serializing/deserializing in a wrong way, or if the spark code generation makes different assumptions. I don't know the relation between the two well enough. Any clarification from your side would be much appreciated.

We used frameless 0.5.0 and spark 2.2 with the same datasets and operations on them and had correct results. And `sizeInBytes` never had a corrupt value as we see with 0.6.0 and spark 2.2.


Here is a snippet of some generated code for our specific case class with the `If` construct below leading to the faulty access to the `UnsafeRow`:
```java

      public java.lang.Object generate(Object[] references) {
        return new SpecificSafeProjection(references);
      }

      class SpecificSafeProjection extends org.apache.spark.sql.catalyst.expressions.codegen.BaseProjection {

        private Object[] references;
        private InternalRow mutableRow;
        private boolean resultIsNull;
private long argValue;
private int argValue1;
private long argValue2;
private long argValue3;
private boolean globalIsNull;

        public SpecificSafeProjection(Object[] references) {
          this.references = references;
          mutableRow = (InternalRow) references[references.length - 1];
          

        }

        public void initialize(int partitionIndex) {
          
        }

        public java.lang.Object apply(java.lang.Object _i) {
          InternalRow i = (InternalRow) _i;
          
com.dcmn.db.BrandTrafficClassification value9 = If(i);
            if (globalIsNull) {
              mutableRow.setNullAt(0);
            } else {
              
              mutableRow.update(0, value9);
            }
          
          return mutableRow;
        }

        
private com.dcmn.db.BrandTrafficClassification If(InternalRow i) {
  InternalRow value2 = i.getStruct(0, 4);
boolean isNull = false;
com.dcmn.db.BrandTrafficClassification value = null;
if (!false && false) {
  
final com.dcmn.db.BrandTrafficClassification value3 = null;
  isNull = true;
  value = value3;
} else {
  
resultIsNull = false;
          if (!resultIsNull) {
            
boolean isNull5 = i.isNullAt(0);
long value5 = isNull5 ? -1L : (i.getLong(0));
            resultIsNull = isNull5;
            argValue = value5;
          }
        
          if (!resultIsNull) {
            
int value6 = i.getInt(1);
            
            argValue1 = value6;
          }
        
          if (!resultIsNull) {
            
long value7 = i.getLong(2);
            
            argValue2 = value7;
          }
        
          if (!resultIsNull) {
            
long value8 = i.getLong(3);
            
            argValue3 = value8;
          }
        
      
      final com.dcmn.db.BrandTrafficClassification value4 = resultIsNull ? null : new com.dcmn.db.BrandTrafficClassification(argValue, argValue1, argValue2, argValue3);
  isNull = resultIsNull;
  value = value4;
}
  globalIsNull = isNull;
  return value;
}
           
      }
    
```

Here is an exampe stacktrace leading to the assertion (with some generated code in between which is equal to the code listing above):

```
[info]   java.lang.AssertionError: sizeInBytes (13500) should be a multiple of 8
[info]   at org.apache.spark.sql.catalyst.expressions.UnsafeRow.pointTo(UnsafeRow.java:166)
[info]   at org.apache.spark.sql.catalyst.expressions.UnsafeRow.getStruct(UnsafeRow.java:462)
[info]   at org.apache.spark.sql.catalyst.expressions.UnsafeRow.getStruct(UnsafeRow.java:62)
[info]   at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificSafeProjection.If$(Unknown Source)
[info]   at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificSafeProjection.apply(Unknown Source)
[info]   at org.apache.spark.sql.Dataset$$anonfun$org$apache$spark$sql$Dataset$$collectFromPlan$1.apply(Dataset.scala:3275)
[info]   at org.apache.spark.sql.Dataset$$anonfun$org$apache$spark$sql$Dataset$$collectFromPlan$1.apply(Dataset.scala:3272)
[info]   at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
[info]   at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
[info]   at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
[info]   at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
[info]   at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
[info]   at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:186)
[info]   at org.apache.spark.sql.Dataset.org$apache$spark$sql$Dataset$$collectFromPlan(Dataset.scala:3272)
[info]   at org.apache.spark.sql.Dataset$$anonfun$collect$1.apply(Dataset.scala:2722)
[info]   at org.apache.spark.sql.Dataset$$anonfun$collect$1.apply(Dataset.scala:2722)
[info]   at org.apache.spark.sql.Dataset$$anonfun$52.apply(Dataset.scala:3253)
[info]   at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:77)
[info]   at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3252)
[info]   at org.apache.spark.sql.Dataset.collect(Dataset.scala:2722)
[info]   at frameless.TypedDataset$$anonfun$collect$1.apply(TypedDataset.scala:291)
[info]   at frameless.TypedDataset$$anonfun$collect$1.apply(TypedDataset.scala:291)
[info]   at frameless.Job$$anon$4.run(Job.scala:38)

```

I hope this information helps.

--- a/build.sbt
+++ b/build.sbt
@@ -90,7 +90,7 @@ lazy val framelessSettings = Seq(
     "com.chuusai" %% "shapeless" % shapeless,
     "org.scalatest" %% "scalatest" % scalatest % "test",
     "org.scalacheck" %% "scalacheck" % scalacheck % "test"),
-  javaOptions in Test ++= Seq("-Xmx1G"),
+  javaOptions in Test ++= Seq("-Xmx1G", "-ea"),
   fork in Test := true,
   parallelExecution in Test := false
 )

--- a/dataset/src/main/scala/frameless/RecordEncoder.scala
+++ b/dataset/src/main/scala/frameless/RecordEncoder.scala
@@ -171,9 +171,14 @@ class RecordEncoder[F, G <: HList, H <: HList]
         field.encoder.fromCatalyst(fieldPath)
       }
 
-      val nullExpr = Literal.create(null, jvmRepr)
       val newArgs = newInstanceExprs.value.from(exprs)
       val newExpr = NewInstance(classTag.runtimeClass, newArgs, jvmRepr, propagateNull = true)
-      If(IsNull(path), nullExpr, newExpr)
+      path match {
+        case BoundReference(0, _, _) => newExpr
+        case _ => {
+          val nullExpr = Literal.create(null, jvmRepr)
+          If(IsNull(path), nullExpr, newExpr)
+        }
+      }
     }
 }

